### PR TITLE
DEV: Redis in VS Code docs

### PR DESCRIPTION
**Pull request by dwdougherty**
This is an early draft of documentation for the **Redis for VS Code** extension. There's still some missing information, such as:

- How the extension will appear in the VS Code marketplace
- how to install it from there

These items are marked **[DWD]...[DWD]** in the doc. Feel free to review for general layout, etc.

I updated the PR to move the new article into the correct place, and I moved all the screenshots to /static/images.

### CODE REVIEW

**Comment by andy-stark-redis** [comment]
@dwdougherty Just to check - this is in the RI section of the docs. Is this intended to be a spin-off of RI rather than a separate product?

File: content/develop/connect/insight/vscode-extension.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+### Connection tools
+
+There are several connection-related tools that are displayed for each opened connection.
+
+<img src="../images/ri-vscode-cnx-tools.png">
+
+Left to right, they are:
+
+- Refresh connection, which retrieves fresh data from the connected Redis database.
+- Edit connection. A dialog similar to the one shown in the previous section is displayed.
```
Comment (andy-stark-redis) [inline content/develop/connect/insight/vscode-extension.md:67]: Maybe keep similar style to the first item?
```suggestion
- Edit connection, which shows a dialog similar to the one described in [Connect to Redis Databases](#connect-db) above.
```

File: content/develop/connect/insight/vscode-extension.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+- Delete connection.
+- Open CLI. [More on this later](#cli).
```
Comment (andy-stark-redis) [inline content/develop/connect/insight/vscode-extension.md:69]:
```suggestion
- Open CLI. See [CLI tool](#cli) below for more information.
```

File: content/develop/connect/insight/vscode-extension.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+
+
+<img src="../images/ri-vscode-cnx-view.png" >
+
+## Key view
+
+Here's what you'll see when there are no keys in your database (the image on the left) and when keys are present (the image on the right).
+
+| View with no keys | View with JSON keys |
```
Comment (andy-stark-redis) [inline content/develop/connect/insight/vscode-extension.md:82]: Maybe do this so the tops of the images align? Having them side-by-side is a good idea, though. BTW, if you use https://www.drawio.com/ to compose the images, there is a [VSCode extension](https://marketplace.visualstudio.com/items?itemName=hediet.vscode-drawio) that lets you edit the resulting SVG image directly in the editor window. Handy for quick updates.

File: content/develop/connect/insight/vscode-extension.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+|:---               |:---                     |
+| <img src="../images/ri-vscode-no-keys.png"> | <img src="../images/ri-vscode-w-keys.png" > |
+
+Redis for VS Code will automatically group the keys based on the one available setting **Delimiter to separate namespaces**, which you can view by clicking on the gear icon in the top-right of the left side pane. Click on the current value to change it. The default setting is the colon (`:`) character.
```
Comment (andy-stark-redis) [inline content/develop/connect/insight/vscode-extension.md:86]: Sorry if too picky! :-) 
```suggestion
Redis for VS Code will automatically group the keys based on the one available setting, **Delimiter to separate namespaces**, which you can view by clicking on the gear icon in the top-right of the left side pane. Click on the current value to change it. The default setting is the colon (`:`) character.
```

File: content/develop/connect/insight/vscode-extension.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+Click on a key to display its contents.
+
+<img src="../images/ri-vscode-key-view.png">
+
+### Key editing tools
+
+There are several editing tools that you can use to edit key data. Each data type will have their own editing capabilities. JSON data is shown in the following examples.
+
+Changes to keys are immediately written to the server.
```
Comment (andy-stark-redis) [inline content/develop/connect/insight/vscode-extension.md:98]:
```suggestion
There are several editing tools that you can use to edit key data. Each data type has its own editing capabilities. The following examples show edits to JSON data. Note that changes to keys are immediately written to the server.
```

File: content/develop/connect/insight/vscode-extension.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+- **Set time-to-live (TTL)**. Click on the **TTL** field to set the duration in seconds.
+
+<img src="../images/ri-vscode-edit-ttl.png">
+
+- **Delete**. Click on the trash can icons to delete the entire key (highlighted in red) or portions of a key (highlighted in yellow).
+
+<img src="../images/ri-vscode-edit-del.png">
+
+- **Add to key**. Click on the `+` button next to the closing bracket (shown highlighted in green above) add a new component to a key.
```
Comment (andy-stark-redis) [inline content/develop/connect/insight/vscode-extension.md:112]:
```suggestion
- **Add to key**. Click on the `+` button next to the closing bracket (shown highlighted in green above) to add a new component to a key. 
```

File: content/develop/connect/insight/vscode-extension.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+|:---               |:---                     |
+| <img src="../images/ri-vscode-recycle-before.png"> | <img src="../images/ri-vscode-recycle-after.png" > |
+
+## The CLI tool {#cli}
+
+The connection tool with the boxed `>_` icon opens a Redis CLI window in the **REDIS CLI** tab at the bottom of the primary pane.
+
+<img src="../images/ri-vscode-cli.png">
+
+The CLI interface works just like the `redis-cli` command.
```
Comment (andy-stark-redis) [inline content/develop/connect/insight/vscode-extension.md:128]: Just for good measure.
```suggestion
The CLI interface works just like the [`redis-cli`]({{< relref "/develop/connect/cli" >}}) command.
```

**Review by andy-stark-redis — COMMENTED** [review]
Few suggestion here and there but nothing major. Probably a bit too early to approve but this is already looking really good.

File: content/develop/connect/insight/vscode-extension.md (hunk 3 above)
Comment (dwdougherty) [reply content/develop/connect/insight/vscode-extension.md:82]: Another great idea. No idea how to proceed.

File: content/develop/connect/redis-for-vscode.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+The Redis for VS Code extension will automatically update just like any other VS Code extension.
+
+## Connect to Redis databases {#connect-db}
+
+Click on the Redis mark (the cursive **R**) in the VS Code menu to begin connecting a Redis database to VS Code.
+
+{{< image filename="images/dev/connect/vscode/vscode-initial.png" >}}
+
+Click on the **+ Connect database** button. A dialog will display in the main pane. In the image shown below, all the options have been checked to show the available details for each connection. These connection details are similar to those accessible from [`redis-cli`]({{< relref "/develop/connect/cli" >}}).
```
Comment (ViktarStarastsenka) [inline content/develop/connect/redis-for-vscode.md:48]: Not sure if it is worth mentioning or not, but if they have local databases (though not OSS cluster) without passwords and with default usernames, then Redis for VS Code will automatically add those databases.

File: content/develop/connect/redis-for-vscode.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+{{< note >}}
+There is no way to change the logical database once you have selected it. If you need to connect to a different logical database, you need to add a separate database connection.
```
Comment (ViktarStarastsenka) [inline content/develop/connect/redis-for-vscode.md:51]: Let's mention that currently, they can't change the logical databases, but we will add this in the upcoming versions.
This is important for our community.

File: content/develop/connect/redis-for-vscode.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+{{< image filename="images/dev/connect/vscode/vscode-settings.png" >}}
+
+Click on a key to display its contents.
+
+{{< image filename="images/dev/connect/vscode/vscode-key-view.png" >}}
+
+### Key editing tools
+
+There are several editing tools that you can use to edit key data. Each data type has its own editing capabilities. The following examples show edits to JSON data. Note that changes to keys are immediately written to the server.
```
Comment (ViktarStarastsenka) [inline content/develop/connect/redis-for-vscode.md:94]: is it worth adding one more example with hashes?
Redis for VS Code supports hash field expiration, so this is a chance to promote Redis 7.4.

Comment (ViktarStarastsenka) [reply content/develop/connect/redis-for-vscode.md:94]: also for hashes we can mention value formatters - a very useful feature (not available for JSONs)

File: content/develop/connect/redis-for-vscode.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+Click on the **+ Connect database** button. A dialog will display in the main pane. In the image shown below, all the options have been checked to show the available details for each connection. These connection details are similar to those accessible from [`redis-cli`]({{< relref "/develop/connect/cli" >}}).
+
+{{< note >}}
+There is no way to change the logical database once you have selected it. If you need to connect to a different logical database, you need to add a separate database connection.
+{{< /note >}}
+
+{{< image filename="images/dev/connect/vscode/vscode-add-menu.png" >}}
+
+After filling out the necessary fields, click on the **Add Redis database** button. The pane on the left side, where you would normally see the Explorer view, shows your database connections.
```
Comment (ViktarStarastsenka) [inline content/develop/connect/redis-for-vscode.md:56]: let's add a note that if they do not have a Redis db, they can create a free Cloud db or use Redis on Docker?

File: content/develop/connect/redis-for-vscode.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+- clients
+description: Connect to Redis from Visual Studio Code.
+linkTitle: Redis for VS Code
+stack: true
+title: Redis for VS Code
+weight: 5
+---
+
+Redis for VS Code is an extension that allows you to connect to your Redis databases from within Microsoft Visual Studio Code.
+After connecting to a database, you can view, add, modify, and delete keys, and interact with your Redis databases using a built-in CLI interface.
```
[the remark this answers is not in the document]
Comment (dwdougherty) [reply content/develop/connect/redis-for-vscode.md:20]: Done. Look for next merge.

File: content/develop/connect/redis-for-vscode.md (hunk 8 above)
Comment (dwdougherty) [reply content/develop/connect/redis-for-vscode.md:48]: Ha! I was wondering why that happened. :) Added. Look for next merge.

File: content/develop/connect/redis-for-vscode.md (hunk 9 above)
Comment (dwdougherty) [reply content/develop/connect/redis-for-vscode.md:51]: We document the products we have today, not the products we'll have in the future. In other words, forward-looking statements are a strict no-no.

File: content/develop/connect/redis-for-vscode.md (hunk 11 above)
Comment (dwdougherty) [reply content/develop/connect/redis-for-vscode.md:56]: This is made plain in the VS Code interface itself. No reason to document it again.

File: content/develop/connect/redis-for-vscode.md (hunk 10 above)
Comment (dwdougherty) [reply content/develop/connect/redis-for-vscode.md:94]: Will do. Look for next merge.

**Review by andy-stark-redis — APPROVED** [review]
Looks good so far :-)

File: content/develop/connect/redis-for-vscode.md (hunk 9 above)
Comment (ViktarStarastsenka) [reply content/develop/connect/redis-for-vscode.md:51]: What if to use something like this? :)
"There is no way to change the logical database once you have selected in the first release of Redis for VS Code"

File: content/develop/connect/redis-for-vscode.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+{{< image filename="images/dev/connect/vscode/vscode-edit-add.png" >}}
+
+- **Refresh**. Click on the refresh icon (the circular arrow) to retrieve fresh data from the server. In the examples below, refresh was clicked and the key now has a new field called "test" that was added by another Redis client.
+
+| Before refresh | After refresh |
+|:---               |:---                     |
+| {{< image filename="images/dev/connect/vscode/vscode-recycle-before.png" >}} | {{< image filename="images/dev/connect/vscode/vscode-recycle-after.png" >}} |
+
+For Hash and String keys, the extension supports numerous value formatters (highlighted in red in the image below). They are:
```
[the remark this answers is not in the document]
Comment (dwdougherty) [reply content/develop/connect/redis-for-vscode.md]: Fixed in next merge.

File: content/develop/connect/redis-for-vscode.md (hunk 9 above)
Comment (dwdougherty) [reply content/develop/connect/redis-for-vscode.md]: I'll modify to something close in the next push.

**Review by ViktarStarastsenka — APPROVED** [review]
LGTM

File: content/develop/connect/insight/vscode-extension.md (hunk 3 above)
Comment (dwdougherty) [reply content/develop/connect/insight/vscode-extension.md]: I got this sorted. Look for next merge.

**Comment by dwdougherty** [comment]
Thank you, @andy-stark-redis, for merging your modified image!

**Pull request merged**